sqlmesh/core/config/loader.py

```diff
@@ -13,53 +13,73 @@
 
 
 def load_config_from_paths(
-    *paths: Path,
+    project_paths: t.List[Path] = [],
+    personal_paths: t.List[Path] = [],
     config_name: str = "config",
     load_from_env: bool = True,
 ) -> Config:
-
     visited_folders: t.Set[Path] = set()
-    python_config: t.Optional[Config] = None
-    non_python_configs = []
-    for path in paths:
-        if not path.exists():
-            continue
-
-        if not path.is_file():
-            raise ConfigError(f"Path '{path}' must be a file.")
-
-        parent_path = path.parent
-        if parent_path in visited_folders:
-            raise ConfigError(f"Multiple configuration files found in folder '{parent_path}'.")
-        visited_folders.add(parent_path)
-
-        extension = path.name.split(".")[-1].lower()
-        if extension in ("yml", "yaml"):
-            if config_name != "config":
+
+    def _load_configs(paths: t.List[Path]) -> t.Dict[str, t.Union[None, Config, t.List]]:
+        python_config: Config | None = None
+        non_python_configs: t.List[t.Dict] = []
+        for path in paths:
+            if not path.exists():
+                continue
+
+            if not path.is_file():
+                raise ConfigError(f"Path '{path}' must be a file.")
+
+            parent_path = path.parent
+            if parent_path in visited_folders:
+                raise ConfigError(f"Multiple configuration files found in folder '{parent_path}'.")
+            visited_folders.add(parent_path)
+
+            extension = path.name.split(".")[-1].lower()
+            if extension in ("yml", "yaml"):
+                if config_name != "config":
+                    raise ConfigError(
+                        "YAML configs do not support multiple configs. Use Python instead."
+                    )
+                non_python_configs.append(load_config_from_yaml(path))
+            elif extension == "py":
+                python_config = load_config_from_python_module(path, config_name=config_name)
+            else:
                 raise ConfigError(
-                    "YAML configs do not support multiple configs. Use Python instead."
+                    f"Unsupported config file extension '{extension}' in config file '{path}'."
                 )
-            non_python_configs.append(load_config_from_yaml(path))
-        elif extension == "py":
-            python_config = load_config_from_python_module(path, config_name=config_name)
-        else:
-            raise ConfigError(
-                f"Unsupported config file extension '{extension}' in config file '{path}'."
-            )
+
+        return {"python_config": python_config, "non_python_configs": non_python_configs}
+
+    project_config_dicts = _load_configs(project_paths)
+    personal_config_dicts = _load_configs(personal_paths)
+
+    no_dialect_err_msg = "Default model SQL dialect is a required configuration parameter - set it in the `model_defaults` key."
```

Comment:
Can this be changed to a validator now on Config and then remove this concept of project config and personal config?

Reply:
I was hoping so, but both the Config [`update_with` method](https://github.com/TobikoData/sqlmesh/blob/eda2a7649827fe395908fb940a456fd232dcff58/sqlmesh/core/config/base.py#L112) and the [`update_field` method](https://github.com/TobikoData/sqlmesh/blob/eda2a7649827fe395908fb940a456fd232dcff58/sqlmesh/core/config/base.py#L81) it calls depend on the ability to instantiate Config objects without all the fields being specified (as would be required with a validator).

My sense is that modifying those will be more complex (in the sense of having side/downstream effects) than this project/non-project approach.

Alternatively, if we're willing to accept some edge cases (like someone hand-constructing project configs instead of using `init`), we can go back to the very first approach of just loading/collating all the configs and checking for a default dialect after that in Context.